Fill out plugin dev chapter more (#2586)
jenkins-infra/jenkins.io: content/doc/developer/plugin-development/optional-dependencies.adoc

You can mark your extension as an optional extension by changing the `@Extension` annotation to `@Extension(optional = true)` and Jenkins doesn't log any class loading errors when reading it.

[source,java]
----
@Extension(optional = true)
public static class YourDescriptor extends Descriptor<YourDescribable> {
    @Override
    public String getDisplayName() {
        return "YourDisplayName";
    }
}
----

To determine if an optional dependency is installed, the Jenkins.getPlugin(String) method can be used.

[source,java]
----
if (Jenkins.getInstance().getPlugin("javanet-uploader") != null) {
    // use classes in the "javanet-uploader" plugin
}
----

## Plugin statuses and behaviors

|===
| Status | getPlugin("dependee") != null | getPlugin("dependee") .getWrapper().isEnabled() | Load class from your plugin

| Not installed | false | NPE | Fail
| Installed without restart | true | true | Succeed
| Installed without restart before your plugin | true | true | Succeed
| Installed | true | true | Succeed
| Uninstalled - not restarted | true | true | Succeed
| Disabled - not restarted | true | false | Succeed
| Disabled | false | NPE | Fail

|===

## Code causes class loading

> When your plugin optionally depends on other plugins, you have to access classes in those plugins only when you have checked those plugins are installed.

Reviewer: ```suggestion
When your plugin optionally depends on another plugin, you have to access classes in the other plugin only when you have checked the plugin is installed.
```

> In most cases, you can do that by testing plugin installations before executing codes that access those classes, as described in Dependencies among plugins#Optional dependencies.

Reviewer: The `plugins#Optional` in this line will not resolve as a link and will not be understood by the reader.

```suggestion
In most cases, you can do that by testing the plugin is installed before executing code that accesses optionally loaded plugin classes.
```

> However, there are cases your class causes loading other classes not when executing codes that access them, but when someone instantiates your class, or calls reflection methods for your class. Especially `Descriptor#<init>` calls `Class#getMethod("getDescriptor")` and it often causes those cases.

Reviewer: ```suggestion
However,there are cases when your class causes loading of other classes at class instantiation or through reflection calls.
For example, `Descriptor#<init>` calls `Class#getMethod("getDescriptor")` and often causes those cases.
```

### Example case

image::/images/developer/optional-dependency-example.png[]

* dependee-plugin and depender-plugin both depend on base-plugin.
* depender-plugin optionally depends on dependee-plugin.
* class Derived in dependee-plugin extends class Base in base-plugin.

## Contained in public method declarations

* Classes in public method declarations will be loaded.

> * Any of following public method declarations cause Derived loaded when SomeClass is instantiated, or someone calls `SomeClass#getMethod` and so on. 

Reviewer: ```suggestion
* Any of following public method declarations cause Derived to be loaded when SomeClass is instantiated, or someone calls `SomeClass#getMethod` and so on. 
```

If dependee-plugin is not installed, it causes `NoClassDefFoundError` even though `doSomething1()` and `doSomething2` are never executed.

[source,java]
----
public class SomeClass {
  // They cause loading Derived even without execution.
   
  public Derived doSomething1() {
    // dosomething...
  }
   
  public void doSomething2(Derived arg0) {
    // dosomething...
  }
}
----

> * You'd better not to use classes in optional depended plugins in any method declarations.

Reviewer: ```suggestion
* It is best to not use classes from optional plugins in method declarations.
```
